substrate/frame/staking-async/runtimes/parachain/src/staking.rs

```diff
@@ -303,11 +305,89 @@ use pallet_staking_async_rc_client as rc_client;
 use xcm::latest::{prelude::*, SendXcm};
 
 pub struct XcmToRelayChain<T: SendXcm>(PhantomData<T>);
-impl<T: SendXcm> rc_client::SendToRelayChain for XcmToRelayChain<T> {
-	type AccountId = AccountId;
 
-	/// Send a new validator set report to relay chain.
-	fn validator_set(report: rc_client::ValidatorSetReport<Self::AccountId>) {
+impl<T: SendXcm> XcmToRelayChain<T> {
+	/// Splits a message until it can pas the validation step of `send_xcm`.
+	///
+	/// It consumes a `ValidatorSetReport`, which should be in full, and possibly splits it into a
+	/// splitter vector thereof, returned as `Ok(vec)`. It also converts to results into
+	/// ready-to-send XCM messages.
+	///
+	/// The maximum number of steps taken is optionally limited by `maybe_max_steps`.
+	///
+	/// If validating still fails, due to any other error not taken into account, it return
+	/// `Err(reason)`.
+	///
+	/// Notes: This is a UMP. Current values are:
+	///
+	/// Polkadot: 65531 (64k)
+	/// Kusama: 65531 (64k)
+	/// Westend: 8388608 (8MB)
+	///
+	///
+	/// To test, this in the relay runtime's genesis config, tweak the `max_downward_message_size`
+	/// and `max_upward_message_size` values.
+	///
+	/// TODO: good for now, but can be refactored and reused both in rc and in AH. What we need is:
+	///
+	/// 1. `trait Splittable` over all types that can be sent over XCM and might be big
+	/// 2. `struct StakingXcmSender<T: SendXcm, S: Splittable>>`
+	fn split_until_validated(
+		report: rc_client::ValidatorSetReport<AccountId>,
+		destination: &Location,
+		maybe_max_steps: Option<u32>,
+	) -> Result<Vec<Xcm<()>>, SendError> {
+		let mut chunk_size = report.new_validator_set.len();
+		let mut steps = 0;
+
+		loop {
+			let current_reports = report.clone().split(chunk_size);
```

Comment:
there's a bunch of unnecessary cloning of vecs happening in this loop, maybe add a TODO for a good-first-issue to come back and refactor this to work with slices or dummy data (dummy data is still good to validate the right chunking size) and avoid copying the real data back and forth multiple times

Reply:
when re-written as described in https://github.com/paritytech/polkadot-sdk/pull/8422/files#diff-a4a0fc985067e6027a30aa3f4c199456c11092f764b7c3313c3f7d26ec762308R331, clones can also be improved.